src/libopensc/card-masktech.c

```diff
@@ -0,0 +1,480 @@
+                                     const sc_security_env_t *env,
+                                     int se_num)
+{
+	struct masktech_private_data *private_data;
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "masktech_set_security_env(), keyRef = 0x%0x, algo = 0x%0x\n",
+		 *env->key_ref, env->algorithm_flags);
+
+	private_data = (struct masktech_private_data *) card->drv_data;
+	if (!private_data)
+		return SC_ERROR_INTERNAL;
+
+	/* save the key reference */
+	if (env->flags & SC_SEC_ENV_KEY_REF_PRESENT) {
+		if (env->key_ref_len != 1) {
+			sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "Invalid key reference supplied.\n");
+			return SC_ERROR_NOT_SUPPORTED;
+		}
+		private_data->rsa_key_ref = env->key_ref[0];
+	}
+
+	return iso_ops->set_security_env(card, env, se_num);
+}
+
+static int masktech_compute_signature(sc_card_t *card,
+                                      const u8 * data,
+                                      size_t datalen,
+                                      u8 * out,
+                                      size_t outlen)
+{
+
+	struct masktech_private_data *private_data;
+	u8 sha256hash[32];
+	static const u8 hdr_sha256[] = {
+		0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60, 0x86, 0x48, 0x01, 0x65,
+		0x03, 0x04, 0x02, 0x01, 0x05, 0x00, 0x04, 0x20
+	};
+	assert(card != NULL && data != NULL && out != NULL);
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "masktech_compute_signature()\n");
+	
+	/* retrieve the key reference */
+	private_data = (struct masktech_private_data *) card->drv_data;
+	if (!private_data)
+		return SC_ERROR_INTERNAL;
+
+	if (private_data->rsa_key_ref == 0x88)
+	{
+		/* for this key reference, the card supports only SHA256 hash and the hash is computed using a digest info */
+		/* check that it is a SHA256 with digest info*/
+		if ((datalen != sizeof(hdr_sha256) + 32) || (memcmp(hdr_sha256, data, sizeof(hdr_sha256)) != 0))
+		{
+			sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "It is not a SHA256 with digestinfo\n");
+			return SC_ERROR_NOT_SUPPORTED;
+		}
+		/* extract the SHA-256 hash */
+		memcpy(sha256hash, (u8 *)(data+(datalen-32)), 32);//only last 32 byte => sha256
+		/* default ISO 7816 functions */
+		return iso_ops->compute_signature(card, sha256hash, 32, out, outlen);
+	}
+	else
+	{
+		/* default ISO 7816 functions */
+		return iso_ops->compute_signature(card, data, datalen, out, outlen);
+	}
+}
+
+static int masktech_decipher(sc_card_t *card,
+                             const u8 * crgram,
+                             size_t crgram_len,
+                             u8 * out,
+                             size_t outlen)
+{
+	int r, cse;
+	sc_apdu_t apdu;
+	u8 rbuf[SC_MAX_EXT_APDU_BUFFER_SIZE];
+	u8 sbuf[SC_MAX_EXT_APDU_BUFFER_SIZE];
+	unsigned char use_ext;
+
+	assert(card != NULL && crgram != NULL && out != NULL);
+	sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "masktech_decipher()\n");
+	
+	use_ext = crgram_len > 255; 
+	cse = use_ext ? SC_APDU_CASE_4_EXT : SC_APDU_CASE_4_SHORT;
```

Comment:
whether the card supports extended length or not, should be decided in it's init function. This allows setting cse `SC_APDU_CASE_4` which lets the lower level decide about extended or short length.